### DISCUSSION TITLE
seaborn.objects: Text isn't shown properly when too close the axis border

### DISCUSSION THREAD

**Discussion by JanPalasek** [mwaskom/seaborn#3170] [Q&A]
Example:

![image](https://user-images.githubusercontent.com/26582151/202478053-7a0e30f9-5321-481c-9a91-ce69bcd18776.png)

Similar issues occur when halign is set for other values. When one bar is too close to the border, this type of problem occurs.

**Comment by mwaskom** [comment]
Which text are you talking about here? And what is the expected behavior?

**Comment by mwaskom** [comment]
Also please put the example code in the issue as text (i.e. something that can be copy-pasted) rather than a screenshot.

**Comment by JanPalasek** [comment]
```python
gd = pd.DataFrame({
    "price": [10000, 0],
    "name": ["A", "B"]
})

(
    so.Plot(data=gd, x="price", y="name", text="price")
    .add(so.Bar())
    .add(so.Text())
)
```

The text that was added by `so.Text()` in the case of B is partially located out of the figure (the 0 value). If I use `so.Text(halign="left")`, then the text in case of A is located out of the figure.

**Comment by mwaskom** [comment]
I'm still interested in what precisely you think the expected behavior is, because two possible things could happen so that the text is not located "outside of the figure":

1. the axis limits would change
2. text that overfills the axes space would remain visible

Neither is exactly simple. (1) is controlled by matplotlib; I don't think seaborn can work around it, and there are some technical reasons why matplotlib's behavior is understandable (or, at least, avoids some significant complexity; https://github.com/matplotlib/matplotlib/issues/10497).

(2) makes sense when text is just slightly running over the axes, but it leads to very surprising behavior in other cases (https://github.com/matplotlib/matplotlib/issues/14354).

You can activate the solution for (2) yourself:

```
(
    so.Plot(data=gd, x="price", y="name", text="price")
    .add(so.Bar())
    .add(so.Text({"clip_on": False}))
)
```
<img width=500 src="https://user-images.githubusercontent.com/315810/202853054-a4e7e191-a696-43b8-8010-f4e3b2a3ebd3.png" />

I'm open to this being the default behavior (it is for `ax.text`). But I don't think there's a solution that will never produce a surprising result.

**Comment by JanPalasek** [comment]
I think the default value for `clip_on` will be better `False`. It will lead to far less unexpected behaviour, especially when using only seaborn (without drawing text etc in matplotlib underneath).

I would maybe also consider setting default as `halign="left"`. It seems to me that it leads to a much more clear plot, where text isn't drawn over the point (when using `so.Dot` with `so.Text`) or weird looking 0 in case of the `so.Bar`.
Also I believe it is more natural to the human eye, since we read the plots in Z pattern (start from upper left, then upper right and then next line similarly). Imo it's better when we first see the dot (or size of the bar) and then the annotation. Maybe there's an counterexample that I don't see though.

**Comment by mwaskom** [comment]
Well, here's a counterexample:

```python
x = list("ABCEDFG")
y = [i * 100 for i in range(1, 8)]
(
    so.Plot(x, y, text=y)
    .add(so.Bar())
    .add(so.Text(halign="left"))
)
```
<img width=500 src="https://user-images.githubusercontent.com/315810/202871167-ff49620c-1fbd-4eef-9a21-9741ef25c63c.png" />

It's possible that halign/valign defaults could depend on the layer orientation, but that's fairly magical. (Also personally I prefer bar labels to be inside the bar, right- or top-aligned to the edge).

**Comment by JanPalasek** [comment]
> It's possible that halign/valign defaults could depend on the layer orientation, but that's fairly magical. (Also personally I prefer bar labels to be inside the bar, right- or top-aligned to the edge).

Thanks for the counter-example. In such a case, I agree it doesn't make sense to change this default (for halign).
